From 7cc8a91a0fc4621055b0a3400c9625ac282187ac Mon Sep 17 00:00:00 2001
From: Ping Gao <ping.a.gao@intel.com>
Date: Wed, 6 Sep 2017 15:42:48 +0800
Subject: [PATCH 0488/1240] drm/i915/gvt: Skip to compare force-nonpriv
 registers

As force-nonpriv registers are updated by command, skip to compare
them with host cache during the first entire non-context MMIOs
check to avoid failure.
Also, when dom0 in guc submission mode, the RING_MODE_GEN7 register could
be different in guest, so ignore it too.

Change-Id: I92d1b3737639b6ea1a0b919fee073d93b2054419
Signed-off-by: Ping Gao <ping.a.gao@intel.com>
Reviewed-by: Kevin Tian <kevin.tian@intel.com>
Reviewed-by: Singh, Satyeshwar <satyeshwar.singh@intel.com>
Reviewed-by: He, Min <min.he@intel.com>
Reviewed-by: Jiang, Fei <fei.jiang@intel.com>
Reviewed-by: Dong, Eddie <eddie.dong@intel.com>
Tested-by: Dong, Eddie <eddie.dong@intel.com>
---
 drivers/gpu/drm/i915/gvt/cmd_parser.c | 2 +-
 drivers/gpu/drm/i915/gvt/gvt.h        | 1 +
 drivers/gpu/drm/i915/gvt/render.c     | 5 ++++-
 3 files changed, 6 insertions(+), 2 deletions(-)

diff --git a/drivers/gpu/drm/i915/gvt/cmd_parser.c b/drivers/gpu/drm/i915/gvt/cmd_parser.c
index 2014cc8..827a5af 100644
--- a/drivers/gpu/drm/i915/gvt/cmd_parser.c
+++ b/drivers/gpu/drm/i915/gvt/cmd_parser.c
@@ -813,7 +813,7 @@ static bool is_shadowed_mmio(unsigned int offset)
 	return ret;
 }
 
-static inline bool is_force_nonpriv_mmio(unsigned int offset)
+bool is_force_nonpriv_mmio(unsigned int offset)
 {
 	return (offset >= 0x24d0 && offset < 0x2500);
 }
diff --git a/drivers/gpu/drm/i915/gvt/gvt.h b/drivers/gpu/drm/i915/gvt/gvt.h
index da61464..9be1ab1 100644
--- a/drivers/gpu/drm/i915/gvt/gvt.h
+++ b/drivers/gpu/drm/i915/gvt/gvt.h
@@ -649,6 +649,7 @@ static inline void intel_gvt_mmio_set_non_context(
 #include "trace.h"
 
 void intel_gvt_mark_noncontext_mmios(struct intel_gvt *gvt);
+bool is_force_nonpriv_mmio(unsigned int offset);
 
 #include "mpt.h"
 
diff --git a/drivers/gpu/drm/i915/gvt/render.c b/drivers/gpu/drm/i915/gvt/render.c
index ab01402..559706a 100644
--- a/drivers/gpu/drm/i915/gvt/render.c
+++ b/drivers/gpu/drm/i915/gvt/render.c
@@ -457,6 +457,7 @@ static int noncontext_mmio_compare(struct intel_vgpu *vgpu, int ring_id)
 	struct drm_i915_private *dev_priv = vgpu->gvt->dev_priv;
 	struct render_mmio *mmio, *mmio_list;
 	int i, array_size;
+	struct intel_engine_cs *engine = dev_priv->engine[ring_id];
 
 	if (IS_SKYLAKE(dev_priv) || IS_BROXTON(dev_priv)) {
 		mmio_list = gen9_render_mmio_list;
@@ -467,7 +468,9 @@ static int noncontext_mmio_compare(struct intel_vgpu *vgpu, int ring_id)
 	}
 
 	for (i = 0, mmio = mmio_list; i < array_size; i++, mmio++) {
-		if (mmio->ring_id != ring_id || mmio->in_context)
+		if (mmio->ring_id != ring_id || mmio->in_context
+			|| is_force_nonpriv_mmio(mmio->reg.reg)
+			|| mmio->reg.reg == RING_MODE_GEN7(engine).reg)
 			continue;
 
 		if (MMIO_COMPARE(vgpu, mmio->reg.reg, mmio->mask))
-- 
2.7.4

